torch/lib/THD/master_worker/master/generic/THDTensorMath.cpp

```diff
@@ -0,0 +1,332 @@
+  return THDTensor_(nElement)(t);
+}
+
+void THDTensor_(diag)(THDTensor *r_, THDTensor *t, int k) {
+  THArgCheck(THDTensor_(nDimension)(t) == 1 || THDTensor_(nDimension)(t) == 2,
+      1, "matrix or a vector expected");
+
+  if (THDTensor_(nDimension)(t) == 1) {
+    real *t_data; // = THDTensor_(data)(t); TODO
+    long t_stride_0 = THDTensor_(stride)(t, 0);
+    long t_size = THDTensor_(size)(t, 0);
+    long sz = t_size + (k >= 0 ? k : -k);
+    real *r__data;
+    long r__stride_0;
+    long r__stride_1;
+    long i;
+
+    THDTensor_(resize2d)(r_, sz, sz);
+    // THDTensor_(zero)(r_); TODO
+    // r__data = THDTensor_(data)(r_); TODO
+    r__stride_0 = THDTensor_(stride)(r_, 0);
+    r__stride_1 = THDTensor_(stride)(r_, 1);
+    r__data += (k >= 0 ? k*r__stride_1 : -k*r__stride_0);
+
+    for (i = 0; i < t_size; i++)
+      r__data[i*(r__stride_0+r__stride_1)] = t_data[i*t_stride_0];
+  } else {
+    real *t_data; // = THDTensor_(data)(t); TODO
+    long t_stride_0 = THDTensor_(stride)(t, 0);
+    long t_stride_1 = THDTensor_(stride)(t, 1);
+    long sz;
+    real *r__data;
+    long r__stride_0;
+    long i;
+
+    if (k >= 0)
+      sz = THMin(THDTensor_(size)(t, 0), THDTensor_(size)(t, 1)-k);
+    else
+      sz = THMin(THDTensor_(size)(t, 0)+k, THDTensor_(size)(t, 1));
+    THDTensor_(resize1d)(r_, sz);
+    // r__data = THDTensor_(data)(r_); TODO
+    r__stride_0 = THDTensor_(stride)(r_, 0);
+
+    t_data += (k >= 0 ? k*t_stride_1 : -k*t_stride_0);
+    for (i = 0; i < sz; i++)
+      r__data[i*r__stride_0] = t_data[i*(t_stride_0+t_stride_1)];
+  }
+}
+
+void THDTensor_(eye)(THDTensor *r, long n, long m) {
+  THArgCheck(n > 0, 1, "invalid argument");
+
+  if (m <= 0)
+    m = n;
+
+  THDTensor_(resize2d)(r, n, m);
+  // THDTensor_(zero)(r); TODO
+
+  masterCommandChannel->sendMessage(
+    packMessage(Functions::tensorEye, r, n, m),
+    THDState::s_current_worker
+  );
+}
+
+
+void THDTensor_(range)(THDTensor *r_, accreal xmin,
+                              accreal xmax, accreal step) {
+  
+  THArgCheck(step > 0 || step < 0, 3, "step must be a non-null number");
+  THArgCheck(((step > 0) && (xmax >= xmin)) || ((step < 0) && (xmax <= xmin)),
+              2, "upper bound and larger bound incoherent with step sign");
+
+  ptrdiff_t size = static_cast<ptrdiff_t>((((xmax - xmin) / step) + 1));
+
+  if (THDTensor_(nElement)(r_) != size)
+    THDTensor_(resize1d)(r_, size);
+
+  masterCommandChannel->sendMessage(
+    packMessage(Functions::tensorRange, r_, xmin, xmax, step),
+    THDState::s_current_worker
+  );
+}
+
+// void THDTensor_(randperm)(THDTensor *r_, THDGenerator *_generator, long n) {
+//   throw std::runtime_error("randperm is not available yet");
+//   THArgCheck(n > 0, 1, "must be strictly positive");
+
+//   THDTensor_(resize1d)(r_, n);
+
+// }
+
+void THDTensor_(reshape)(THDTensor *r_, THDTensor *t, THLongStorage *size) {
+  THDTensor_(resize)(r_, size, NULL);
+  //THDTensor_(copy)(r_, t); TODO
```

Comment:
I agree. I didn't mention it at first but I meant something like:

- **in master**: call `THDTensor_(resize)` (which should also send `resize` function to worker); send `reshape` function to worker
- **in worker**: get `resize` function and resize; get `reshape` function and call `THTensor_(reshape)` function which will resize once again (it should do nothing actually because tensor is already resized) and copy data.

So there is no need for `copy` in master.

If I am not mistaken "my" approach is semantically equal to `resize` and `copy` without sending `reshape` function to worker.